ENH: Support MultiIndex columns in parquet (#34777)

hweecat opened the pull request:
- [x] closes #34777
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Support MultiIndex for columns in parquet format by updating value column names check to handle MultiIndexes.

hweecat committed:
ENH: Support MultiIndex columns in parquet (#34777)

1. Update check to handle MultiIndex columns for parquet format

hweecat committed:
ENH: Support MultiIndex columns in parquet GH34777

1. add whatsnew entry

hweecat committed:
Merge remote-tracking branch 'upstream/master' into io-parquet-multiindex

dsaxton commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -297,6 +297,7 @@ I/O
 - :meth:`to_csv` did not support zip compression for binary file object not having a filename (:issue: `35058`)
 - :meth:`to_csv` and :meth:`read_csv` did not honor `compression` and `encoding` for path-like objects that are internally converted to file-like objects (:issue:`35677`, :issue:`26124`, and :issue:`32392`)
 - :meth:`to_picke` and :meth:`read_pickle` did not support compression for file-objects (:issue:`26237`, :issue:`29054`, and :issue:`29570`)
+- :meth:`to_parquet` did not support MultiIndex for columns in parquet format (:issue:`34777`)
```
```suggestion
- :meth:`to_parquet` did not support :class:`MultiIndex` for columns in parquet format (:issue:`34777`)
```

hweecat replied:
> Can you also add a test?

Thanks @dsaxton - adding a test for writing MultiIndexes for columns with string column names (test_write_column_multiindex_string).

dsaxton reviewed (commented):
Can you also add a test?

dsaxton commented on pandas/io/parquet.py:
```diff
@@ -53,9 +53,15 @@ def validate_dataframe(df: DataFrame):
-        if df.columns.inferred_type not in {"string", "empty"}:
-            raise ValueError("parquet must have string column names")
+        # must have value column names for all index levels (strings only)
+        if df.columns.nlevels > 1:
```
If you replace this check with isinstance(df.columns, MultiIndex) that might silence mypy

hweecat replied:
Thanks @dsaxton for the suggestion - looks like replacing the check with `isinstance(df.columns, MultiIndex)` did the trick with silencing mypy on my dev environment (and should do the trick on CI too)!

dsaxton replied:
Great, glad it helped

hweecat committed:
ENH: Support MultiIndex columns in parquet #34777

1. Update check to handle MultiIndex columns for parquet format
2. Edit whatsnew entry.
3. Add test for writing MultiIndex columns with string column names

dsaxton commented on pandas/tests/io/test_parquet.py:
```diff
@@ -410,11 +410,24 @@ def test_write_multiindex(self, pa):
         df = pd.DataFrame(np.random.randn(4, 3), columns=mi_columns)
         self.check_error_on_write(df, engine, ValueError)
 
+    def test_write_column_multiindex_string(self, pa):
```
Can you include the issue number or URL as a comment?

hweecat replied:
Added issue number as a comment on test in latest commit.

hweecat committed:
ENH: Support MultiIndex columns in parquet #34777

1. Include issue number as a comment on added test

jreback commented on pandas/io/parquet.py:
```diff
@@ -53,9 +53,15 @@ def validate_dataframe(df: DataFrame):
+            if not all(
+                x.inferred_type in {"string", "empty"} for x in df.columns.levels
+            ):
+                raise ValueError("parquet must have string column names")
```
can say something about 'for all values in each level of the MultiIndex'

hweecat replied:
Thanks @jreback for the suggestion on the exception statement - adding that into my next commit!

jreback commented on pandas/tests/io/test_parquet.py:
```diff
@@ -410,11 +410,25 @@ def test_write_multiindex(self, pa):
 
+    def test_write_column_multiindex_string(self, pa):
+        # GH #34777
+        # Not supported in fastparquet as of 0.1.3 or older pyarrow version
```
are we > than the min pyarrow version?

hweecat replied:
Based on the min versions listed in pandas dependencies, the min pyarrow version is 0.15 while we are currently at 0.16 - at least for the dev environment that I'm working on.

jreback commented on pandas/tests/io/test_parquet.py:
```diff
@@ -410,11 +410,25 @@ def test_write_multiindex(self, pa):
+            ["bar", "bar", "baz", "baz", "foo", "foo", "qux", "qux"],
+            ["one", "two", "one", "two", "one", "two", "one", "two"],
+        ]
+        df = pd.DataFrame(np.random.randn(8, 8), columns=arrays)
```
can you add names to the MultiIndex levels. do these round trip?

hweecat replied:
After adding names to the MultiIndex levels, looks like they do round trip on pytest.

jreback commented on pandas/io/parquet.py:
```diff
@@ -53,9 +53,15 @@ def validate_dataframe(df: DataFrame):
-            raise ValueError("parquet must have string column names")
+        # must have value column names for all index levels (strings only)
+        if isinstance(df.columns, MultiIndex):
+            if not all(
```
can you add a test for this

hweecat replied:
@jreback Since the enhancement is to add support for MultiIndex columns in parquet and the check is to handle MultiIndex columns in addition to single-level Index columns, I've added both round-trip and error-on-write tests for both MultiIndexes and single-level Indexes for columns in parquet format.

Do let me know if the coverage from the added tests are sufficient based on my latest commit - and whether additional tests are needed. :)

jreback commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -297,6 +297,7 @@ I/O
 - :meth:`to_csv` did not support zip compression for binary file object not having a filename (:issue: `35058`)
 - :meth:`to_csv` and :meth:`read_csv` did not honor `compression` and `encoding` for path-like objects that are internally converted to file-like objects (:issue:`35677`, :issue:`26124`, and :issue:`32392`)
 - :meth:`to_picke` and :meth:`read_pickle` did not support compression for file-objects (:issue:`26237`, :issue:`29054`, and :issue:`29570`)
+- :meth:`to_parquet` did not support :class:`MultiIndex` for columns in parquet format (:issue:`34777`)
```
i would move this to other enhancements; say this is enabled with pyarrow=.....

hweecat replied:
Moved the whatsnew entry to "other enhancements" - thanks!

hweecat committed:
ENH: Support MultiIndex columns in parquet #34777

1. Add tests for writing Indexes and MultiIndexes for columns
2. Edit message for check to handle  MultiIndex columns for parquet
3. Edit whatsnew entry to move entry to other enhancements

pep8speaks commented:
Hello @hweecat! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/io/parquet.py`](https://github.com/pandas-dev/pandas/blob/9e8f4ebd1cba08abd01440e62eb2cd1ac72177f4/pandas/io/parquet.py):

> [Line 62:89](https://github.com/pandas-dev/pandas/blob/9e8f4ebd1cba08abd01440e62eb2cd1ac72177f4/pandas/io/parquet.py#L62): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 88 characters)

hweecat committed:
ENH: Support MultiIndex columns in parquet #34777

1. Fix PEP8 issue for error message in check for MultiIndex columns

hweecat committed:
Merge remote-tracking branch 'upstream/master' into io-parquet-multiindex

hweecat committed:
Merge remote-tracking branch 'upstream/master' into io-parquet-multiindex

hweecat committed:
Merge remote-tracking branch 'upstream/master' into io-parquet-multiindex

hweecat committed:
Merge remote-tracking branch 'upstream/master' into io-parquet-multiindex

jreback reviewed (changes requested):
looks fine, can you add a whatsnew note; enhancements in 1.2

does this have a min pyarrow version?

hweecat committed:
Merge remote-tracking branch 'upstream/master' into io-parquet-multiindex

hweecat committed:
ENH: Support MultiIndex columns in parquet #34777

add whatsnew entry: enhancements in 1.2

hweecat commented:
@jreback Since the enhancements work on pyarrow 0.15.0, we could leave the min pyarrow version as >= 0.15.0 for this pull request.

I've added a whatsnew note on enhancements in 1.2 under "Other enhancements" - feel free to let me know if a more detailed whatsnew note is needed. :)

P.S. Tests are failing after I did a rebase to update my branch.

dsaxton commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -484,7 +485,6 @@ Other
 - Fixed metadata propagation in the :class:`Series.dt` and :class:`Series.str` accessors (:issue:`28283`)
 - Bug in :meth:`Index.union` behaving differently depending on whether operand is a :class:`Index` or other list-like (:issue:`36384`)
 - Passing an array with 2 or more dimensions to the :class:`Series` constructor now raises the more specific ``ValueError``, from a bare ``Exception`` previously (:issue:`35744`)
-
```
Can you add this back? Seems to be causing CI failure.

hweecat replied:
Whoops, accidentally removed that line! Just added it back - hope this fixes the doc failure.

hweecat committed:
fix doc failure

hweecat committed:
Merge remote-tracking branch 'upstream/master' into io-parquet-multiindex

hweecat committed:
Merge remote-tracking branch 'upstream/master' into io-parquet-multiindex

alimcmaster1 commented:
Test failures look unrelated:

```
fsspectest = <pandas.conftest.fsspectest.<locals>.TestMemoryFS object at 0x7fb2b99e5e50>
extension = 'xls'

    @pytest.mark.parametrize("extension", ["xlsx", "xls"])
    def test_excel_options(fsspectest, extension):
        df = DataFrame({"a": [0]})
    
        path = f"testmem://test/test.{extension}"
    
>       df.to_excel(path, storage_options={"test": "write"}, index=False)

pandas/tests/io/test_fsspec.py:133: 
```

hweecat committed:
Merge remote-tracking branch 'upstream/master' into io-parquet-multiindex

charlesdong1991 commented:
/azp run

charlesdong1991 commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -221,6 +221,7 @@ Other enhancements
 -
 - Added methods :meth:`IntegerArray.prod`, :meth:`IntegerArray.min`, and :meth:`IntegerArray.max` (:issue:`33790`)
 - Where possible :meth:`RangeIndex.difference` and :meth:`RangeIndex.symmetric_difference` will return :class:`RangeIndex` instead of :class:`Int64Index` (:issue:`36564`)
+- :meth:`to_parquet` now supports :class:`MultiIndex` for columns in parquet format (:issue:`34777`)
```
```suggestion
- :meth:`DataFrame.to_parquet` now supports :class:`MultiIndex` for columns in parquet format (:issue:`34777`)
```

charlesdong1991 reviewed (approved):
CI is back green, one minor comment on whatsnew

jreback commented:
@hweecat small comments, can you merge master and ping on green.

charlesdong1991 committed:
Update doc/source/whatsnew/v1.2.0.rst

hweecat committed:
Merge remote-tracking branch 'upstream/master' into io-parquet-multiindex

hweecat committed:
Merge remote-tracking branch 'upstream/master' into io-parquet-multiindex

jreback commented:
thanks @hweecat and @charlesdong1991

charlesdong1991 commented:
all credits to @hweecat very nice job!

mraxilus commented:
Any reason why this explicitly raises errors for non-string multi-indicies?
